### PR TITLE
Refactor theory model

### PR DESCRIPTION
**Pull request by ajreynol**
This is work towards #1199.  This commit encapsulates most of the data members of TheoryModel and RepSet (a key utility in the model used for finite model finding for interpreting types).  It also documents RepSet.

Most of the changes were the result of change the accesses to TheoryModel::d_rep_set to TheoryModel::getRepSet() and various access functions were added to RepSet.  

Only very minor changes in actual code behavior were done in this commit.

TODO (as noted in #1199):
- Further encapsulate RepSet
- Make all external references to RepSet const (in other words, external objects should not be able to modify the interpretation of types).
- Document and cleanup RepSetIterator (in src/theory/rep_set.h)

### CODE REVIEW

**Comment by ajreynol** [comment]
This also fixes #970.

File: src/theory/rep_set.h [hunk 1]
```diff
@@ -25,32 +25,72 @@ namespace theory {
...
+   */
+  Node getTermForRepresentative( Node n ) const;
+  /** set term for representative 
+   * Called when t is interpreted as value n. Subsequent class to getTermForRepresentative( n ) will return t.
+   */
+  void setTermForRepresentative( Node n, Node t );
+  /** get existing domain value, with possible exclusions
+    *   This function returns a term in d_type_reps[tn] but not in exclude
+    */
+  Node getDomainValue( TypeNode tn, std::vector< Node >& exclude ) const;
```
Comment (timothy-king) [inline src/theory/rep_set.h]: #include vector

File: src/theory/rep_set.h [hunk 2]
```diff
@@ -25,32 +25,72 @@ namespace theory {
...
   std::map< TypeNode, std::vector< Node > > d_type_reps;
-  std::map< TypeNode, bool > d_type_complete;
-  std::map< Node, int > d_tmap;
-  // map from values to terms they were assigned for
-  std::map< Node, Node > d_values_to_terms;
   /** clear the set */
   void clear();
-  /** has type */
+  /** does this set have representatives of type tn? */
   bool hasType( TypeNode tn ) const { return d_type_reps.find( tn )!=d_type_reps.end(); }
```
Comment (timothy-king) [inline src/theory/rep_set.h]: Include type node

**Comment by ajreynol** [comment]
Tim's comments are addressed in my latest commit.

**Comment by ajreynol** [comment]
Changing this to normal since further refactoring to TheoryModel is soon going to be needed for #1168, where GE wants models for transcendental functions.

File: src/theory/rep_set.h [hunk 3]
```diff
@@ -17,40 +17,83 @@
...
+ * in the equality engine of a model object.  In the following, "representative" means a value
+ * that exists in this set.
+ * 
+ * This class is used for finite model finding and other exhaustive instantiation-based
+ * methods. The class goes beyond just maintaining a list of values that occur in the equality
+ * engine in the following ways:
+ * (1) It maintains a partial mapping from representatives to a term that has that value in the current
+ * model.  This is important because algorithms like the instantiation method in Reynolds et al CADE 2013
+ * act on "term models" where domains in models are interpreted as a set of representative terms. Hence,
+ * instead of instantiating with e.g. uninterpreted constants u, we instantiate with the correspond term
```
Comment (mpreiner) [inline src/theory/rep_set.h]: > with the correspond term

-->

> with the corresponding term

**Comment by mpreiner** [comment]
Can you also make sure that the comments fit into 80 character lines? Apply clang-format?

**Pull request merged**